### PR TITLE
gvstream: improve performance of receiving loop (issue #616)

### PR DESCRIPTION
**Pull request by lyodaom**
See [issue 616](https://github.com/AravisProject/aravis/issues/616)

### CODE REVIEW

**Comment by lyodaom** [comment]
The socket was set to non-blocking mode to allow `g_socket_receive_messages` to return fewer number of messages.
Some tests with arv-fake-gv-camera were done and the improvement was significant.
Before the change:
```bash
$ arv-camera-test-0.8 -w 512 -h 512 -f 1000 --duration 5
Looking for the first available camera
vendor name            = Aravis
model name             = Fake
device serial number   = GV01
image width            = 512
image height           = 512
horizontal binning     = 1
vertical binning       = 1
exposure               = 10000 µs
gain                   = 0 dB
payload                = 262144 bytes
gv n_stream channels   = 1
gv current channel     = 0
gv packet delay        = 0 ns
gv packet size         = 1400 bytes
542 frames/s -     142 MiB/s
549 frames/s -     144 MiB/s - 1 error
552 frames/s -     145 MiB/s
500 frames/s -     131 MiB/s - 15 errors
530 frames/s -     139 MiB/s
n_completed_buffers    = 2673
n_failures             = 16
n_underruns            = 0
n_timeouts             = 0
n_aborteds             = 0
n_missing_frames       = 0
n_size_mismatch_errors = 0
n_received_packets     = 523776
n_missing_packets      = 1445
n_error_packets        = 0
n_ignored_packets      = 0
n_resend_requests      = 0
n_resent_packets       = 0
n_resend_ratio_reached = 0
n_resend_disabled      = 0
n_duplicated_packets   = 0
n_transferred_bytes    = 708415692
n_ignored_bytes        = 0
```

After the change:
```bash
$ arv-camera-test-0.8 -w 512 -h 512 -f 1000 --duration 5
Looking for the first available camera
vendor name            = Aravis
model name             = Fake
device serial number   = GV01
image width            = 512
image height           = 512
horizontal binning     = 1
vertical binning       = 1
exposure               = 10000 µs
gain                   = 0 dB
payload                = 262144 bytes
gv n_stream channels   = 1
gv current channel     = 0
gv packet delay        = 0 ns
gv packet size         = 1400 bytes
714 frames/s -     187 MiB/s
778 frames/s -     204 MiB/s - 5 errors
805 frames/s -     211 MiB/s
911 frames/s -     239 MiB/s
879 frames/s -     230 MiB/s
n_completed_buffers    = 4087
n_failures             = 5
n_underruns            = 0
n_timeouts             = 0
n_aborteds             = 0
n_missing_frames       = 0
n_size_mismatch_errors = 0
n_received_packets     = 797823
n_missing_packets      = 446
n_error_packets        = 0
n_ignored_packets      = 0
n_resend_requests      = 0
n_resent_packets       = 0
n_resend_ratio_reached = 0
n_resend_disabled      = 0
n_duplicated_packets   = 0
n_transferred_bytes    = 1079100544
n_ignored_bytes        = 0
```

**Comment by lyodaom** [comment]
I think it might be ready to consider, I'll try testing it with real hardware though

**Comment by lyodaom** [comment]
This change was successfully tested using a few cameras Mako G-319B (Allied Vision), and it was confirmed to reduce the error rate.

**Comment by lyodaom** [comment]
I don't know why  the windows builds are failing, I replicated the build environment using msys2 and the tests did not fail:
```bash
user@DESKTOP MSYS ~/aravis/build
$ meson test
ninja: Entering directory `C:/msys64/home/user/aravis/build'
ninja: no work to do.
1/9 aravis:main / evaluator      OK              0.09s
2/9 aravis:main / buffer         OK              0.17s
3/9 aravis:main / misc           OK              0.06s
4/9 aravis:main / dom            OK              0.08s
5/9 aravis:main / genicam        OK              0.11s
6/9 aravis:main / fake           OK              1.34s
7/9 aravis:network / fakegv      OK              9.20s
8/9 aravis:python / fake.py      OK              0.95s
9/9 aravis:python / exception.py OK              1.20s


Ok:                 9
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Full log written to C:/msys64/home/user/aravis/build/meson-logs/testlog.txt
```

Maybe the CI was not using latest msys2

**Comment by EmmanuelP** [comment]
> I don't know why the windows tests are failing

I don't know either. Your patch looks good, and the CI pipeline is successful on the main branch. I even tried to rebase your patch over the main branch, but the pipeline still fails.

Unfortunately, this CI failure is a real blocker. I can not leave the Windows build out of the CI. You may want to try to conditionally build without this optimization on the Windows platform.

**Comment by lyodaom** [comment]
I've conditionally set the number of messages because I suspect this could be a resource limit related issue, let's see what happens

**Comment by lyodaom** [comment]
OK, I think that worked: https://github.com/lyodaom/aravis/runs/4730161807?check_suite_focus=true

**Comment by EmmanuelP** [comment]
As a bonus, you could try to use a sensible value for the incoming message buffer size, instead of unconditionally use 64kB.

**Comment by lyodaom** [comment]
OK, I've created a meson parameter called 'gv-buffers', I thought referring to buffers was more meaningful to user (instead of messages).
The parameter end up in file arvparams.h because arvfeatures.h sounds for enable/disable options.

Regarding the packet buffer size, it was calculated based on the GVSP packet size.

File: meson_options.txt [hunk 1]
```diff
@@ -8,6 +8,10 @@ option('packet-socket', type: 'feature', value: 'auto', description : 'Enable pa
 option('tests', type: 'boolean', value: true, description: 'Build tests')
 option('fast-heartbeat', type: 'boolean', value: false, description: 'Enable faster heartbeat rate')
 
+# Parameters
+
+option('gv-buffers', type: 'integer', min: 1, value: 16, description: 'Number of buffers used to receive GVSP packets')
```
Comment (EmmanuelP) [inline meson_options.txt:13]: Please rename the parameter to gv-n-buffers

File: src/arvparams.h.in [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * ARV_GV_STREAM_NUM_BUFFERS
+ *
+ * Number of buffers used to receive GVSP packets
+ *
+ * Since: 0.8.21
+ */
+
+#mesondefine ARV_GV_STREAM_NUM_BUFFERS
+
+#endif
```
Comment (EmmanuelP) [inline src/arvparams.h.in:36]: A dedicated header is a bit overkill, but let's keep it. Rename it to arvparamsprivate.h please, and don't install it. I don't want to make this define public, at least for now.

File: src/arv.h [hunk 3]
```diff
@@ -54,6 +54,7 @@
 #include <arvfakestream.h>
 
 #include <arvfeatures.h>
+#include <arvparams.h>
```
Comment (EmmanuelP) [inline src/arv.h:57]: Don't include the parameter header in arv.h.

**Review by EmmanuelP — CHANGES_REQUESTED** [review]
There is some minor issues to fix, but the patch looks fine.

I still have to find some time to test it with real hardware before I can merge.

File: src/arv.h (hunk 3 above)
Comment (lyodaom) [reply src/arv.h]: Done

File: meson_options.txt (hunk 1 above)
Comment (lyodaom) [reply meson_options.txt]: Done

**Comment by EmmanuelP** [comment]
Thanks a lot.

**Pull request merged**